workshops/09-Polymorphism-Workshop/README.md

```diff
@@ -0,0 +1,151 @@
+# Polymorpism
+
+One fundamental feature of object-oriented programming is _polymorpism_.
+
+Consider the following code:
+
+```ruby
+class AccountingDocument 
+  def initialize(type:, total:)
+    @type = type
+    @total = total
+  end
+
+  def get_journal
+    case @type
+      when :sale
+        [
+          :trade_debtors, -@total,
+          :revenue, @total
+        ] 
+      when :purchase
+        [
+          :trade_creditors, @total,
+          :loss, -@total
+        ]
+      when :expense
+        [
+          :employee_expenses, @total,
+          :bank_account, -@total
+        ]
+      when :draft
+        nil
+      else
+    end
+  end
+end
+```
+
+Where an example usage looks like:
+
+```ruby
+AccountingDocument.new(type: :purchase, total: '100.00')
+```
+
+## The problem
+
+Every time we add a new type of `AccountingDocument` this switch statement will become more complicated.
+
+Not only that - our AccountingDocument class will only ever grow in complexity as we discover more journal types.
+
+## An alternative
+
+Instead, we can use polymorpism:
+
+```ruby
+class AccountingDocument
+  def initialize(journal_disposition:, total:)
+    @journal_disposition = journal_disposition
+    @total = total
+  end
+
+  def get_journal
+    @journal_disposition.get_journal(@total)
+  end
+end
+```
+
+The above class can be said to be "Open for Extension" and "Closed for Modification".
+
+We can then define classes for each "JournalDisposition"
+
+```ruby
+class SaleJournalDisposition
+  def get_journal(total)
+    [
+      :trade_debtors, -@total,
+      :revenue, @total
+    ] 
+  end
+end
+```
+
+```ruby
+class PurchaseJournalDisposition
+  def get_journal(total)
+    [
+      :trade_creditors, @total,
+      :revenue, -@total
+    ] 
+  end
+end
+```
+
+```ruby
+class ExpenseJournalDisposition
+  def get_journal(total)
+    [
+      :employee_expenses, total,
+      :bank_account, -@total
+    ]
+  end
+end
+```
+
+```ruby
+class NoJournalDisposition
+end
+```
+
+But wait! We still need something which can convert `data` into `behaviour`
```

Comment:
Is it worth reinforcing _why_ we need this? I'm not sure if it's been a common lesson across the course.

> ... into `behaviour`, so that we can clearly express our app's intent, through declaring composition.

(I don't think that's a great sentence, but you get me?)